firestore/integration_test_internal/src/jni/arena_ref_test.cc

```diff
@@ -0,0 +1,278 @@
+  Local<String> string1 = env().NewStringUtf("hello world");
+  Local<String> string2 = env().NewStringUtf("hello earth");
+
+  ArenaRef arena_ref1(env(), string1);
+  ArenaRef arena_ref2(env(), string2);
+  arena_ref1 = arena_ref2;
+
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string2));
+  EXPECT_TRUE(env().IsSameObject(arena_ref2.get(env()), string2));
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromNullObjectItself) {
+  ArenaRef arena_ref1;
+  arena_ref1 = arena_ref1;
+  EXPECT_EQ(arena_ref1.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, CopyAssignsFromValidObjectItself) {
+  Local<String> string1 = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1(env(), string1);
+  arena_ref1 = arena_ref1;
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string1));
+}
+
+TEST_F(ArenaRefTestAndroid, MoveConstructsFromNull) {
+  ArenaRef arena_ref1;
+  ArenaRef arena_ref2(std::move(arena_ref1));
+  EXPECT_EQ(arena_ref2.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, MoveConstructsFromValid) {
+  Local<String> string = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref2(env(), string);
+  ArenaRef arena_ref3(std::move(arena_ref2));
+  EXPECT_TRUE(env().IsSameObject(arena_ref3.get(env()), string));
+}
+
+TEST_F(ArenaRefTestAndroid, MoveAssignsFromNullToNull) {
+  ArenaRef arena_ref1, arena_ref2;
+  arena_ref2 = std::move(arena_ref1);
+  EXPECT_EQ(arena_ref2.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, MoveAssignsFromNullToValid) {
+  Local<String> string = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1;
+  ArenaRef arena_ref2(env(), string);
+  arena_ref2 = std::move(arena_ref1);
+  EXPECT_EQ(arena_ref2.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, MoveAssignsFromValidToNull) {
+  Local<String> string = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1;
+  ArenaRef arena_ref2(env(), string);
+  arena_ref1 = std::move(arena_ref2);
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string));
+}
+
+TEST_F(ArenaRefTestAndroid, MoveAssignsFromValidToValid) {
+  Local<String> string1 = env().NewStringUtf("hello world");
+  Local<String> string2 = env().NewStringUtf("hello earth");
+
+  ArenaRef arena_ref1(env(), string1);
+  ArenaRef arena_ref2(env(), string2);
+  arena_ref1 = std::move(arena_ref2);
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string2));
+}
+
+TEST_F(ArenaRefTestAndroid, MoveAssignsFromNullObjectItself) {
+  ArenaRef arena_ref1;
+  arena_ref1 = std::move(arena_ref1);
+  EXPECT_EQ(arena_ref1.get(env()).get(), nullptr);
+}
+
+TEST_F(ArenaRefTestAndroid, MoveAssignsFromValidObjectItself) {
+  Local<String> string1 = env().NewStringUtf("hello world");
+
+  ArenaRef arena_ref1(env(), string1);
+  arena_ref1 = std::move(arena_ref1);
+  EXPECT_TRUE(env().IsSameObject(arena_ref1.get(env()), string1));
+}
+
+TEST_F(ArenaRefTestAndroid, ThrowBeforeConstruct) {
```

Comment:
nit: Can you rename the "ThrowBeforeXXX" test names to "XXXWithPendingException" or something like that? I think using the term "with pending exception" more clearly conveys what scenario the test is testing.